istioctl/cmd/analyze.go

```diff
@@ -308,15 +309,26 @@ istioctl experimental analyze -L
 	return analysisCmd
 }
 
-func gatherFiles(args []string) ([]string, error) {
-	var result []string
-	for _, a := range args {
-		if _, err := os.Stat(a); err != nil {
-			return nil, fmt.Errorf("could not find file %q", a)
+func gatherFiles(args []string) ([]io.Reader, error) {
+	var readers []io.Reader
+	var r *os.File
+	var err error
+	for _, f := range args {
+		if f == "-" {
+			if isatty.IsTerminal(os.Stdin.Fd()) {
+				return nil, fmt.Errorf("No content from stdin to read, but saw '-' arg")
+			}
+			r = os.Stdin
+		} else {
+			r, err = os.Open(f)
+			if err != nil {
+				return nil, err
+			}
+			runtime.SetFinalizer(r, func() { r.Close() })
```

Comment:
Just to make sure I understand correctly: Finalizers get run when the object gets GCed, but there's nothing that guarantees when this would happen.

So, for example, do we need to worry about the following scenario?

Integ test A starts, invokes istioctl analyze (via func call), reads file `Foo`
Integ test A ends. The io.Reader associated with `Foo` is now eligible for GC.
Integ test B starts. It also tries to open `Foo`. <-- What happens here, if we haven't GCed?

Reply:
Assuming the integ tests are different processes then nothing, it's perfectly fine to have two processes open the same file for reading. If they're in the same process I believe that we'll just have two open file descriptors, so even if the finalizer runs then the second one that B is using should be valid. I can't think of an issue here that might cause anything wonky.

Reply:
...toying around with this a bit, it looks like the answer is... not much, besides leaking the FD. I guess this is very unlikely to matter much in practice.